File: src/cfrpy/cfrplus.py
from typing import Dict
import random

from .cfr import matching
from .game import Game, Player, I, A_inv


# not sure about correctness of this implementation
# for one, it is not vector form as the paper describes
# the only differences with external sampling cfr are:
# - the strategy updates are linearly weighted (as in linear CFR)
# - the regrets are clipped above 0 when the strategy is updated
#   (regret matching+ learner)


def eswalkplus(
    game: Game[A_inv, I],
    player: Player,
    regrets: Dict[I, Dict[A_inv, float]],
    strategies: Dict[I, Dict[A_inv, float]],
    t: int,
):
    if game.terminal:
        return game.payoff(player)

    if game.chance:
        action = game.sample()
        return eswalkplus(game.apply(action), player, regrets, strategies, t)

    infoset = game.infoset(game.active)
    actions = infoset.actions()

    if infoset not in regrets:
        regrets[infoset] = {action: 0 for action in actions}
    if infoset not in strategies:
        strategies[infoset] = {action: 0 for action in actions}

    R = regrets[infoset]
    S = strategies[infoset]

    strategy = matching(list(R.values()))

    if game.active != player:
        (action,) = random.choices(actions, weights=strategy)
        value = eswalkplus(game.apply(action), player, regrets, strategies, t)

        for action, p in zip(actions, strategy):
            S[action] += t * p
            R[action] = max(0, R[action])

        return value

    cfs = {action: 0 for action in actions}
    value = 0

    for action, p in zip(actions, strategy):
        cf = eswalkplus(game.apply(action), player, regrets, strategies, t)
        cfs[action] = cf
        value += p * cf

    for action in actions:
        R[action] += cfs[action] - value

    return value

File: src/cfrpy/cfrplus.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <random>
#include <algorithm>
#include "cfr.h"
#include "game.h"

using namespace std;

unordered_map<I, unordered_map<A_inv, float>> eswalkplus(
    Game<A_inv, I> game,
    Player player,
    unordered_map<I, unordered_map<A_inv, float>> regrets,
    unordered_map<I, unordered_map<A_inv, float>> strategies,
    int t)
{
    if (game.terminal)
        return game.payoff(player);

    if (game.chance) {
        auto action = game.sample();
        return eswalkplus(game.apply(action), player, regrets, strategies, t);
    }

    auto infoset = game.infoset(game.active);
    auto actions = infoset.actions();

    if (regrets.count(infoset) == 0) {
        for (auto action : actions)
            regrets[infoset][action] = 0;
    }

    if (strategies.count(infoset) == 0) {
        for (auto action : actions)
            strategies[infoset][action] = 0;
    }

    auto R = regrets[infoset];
    auto S = strategies[infoset];

    auto strategy = matching(R);

    if (game.active != player) {
        auto action = random_element(actions.begin(), actions.end(), strategy);
        auto value = eswalkplus(game.apply(action), player, regrets, strategies, t);

        for (auto action, p : zip(actions, strategy)) {
            S[action] += t * p;
            R[action] = max(0, R[action]);
        }

        return value;
    }

    unordered_map<A_inv, float> cfs;
    float value = 0;

    for (auto action, p : zip(actions, strategy)) {
        auto cf = eswalkplus(game.apply(action), player, regrets, strategies, t);
        cfs[action] = cf;
        value += p * cf;
    }

    for (auto action : actions)
        R[action] += cfs[action] - value;

    return value;
}
